In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#Read the previous application dataset
df1=pd.read_csv('/content/drive/MyDrive/Creditcard/previous_application.csv')
df1.shape

(1670214, 37)

In [ ]:
df1.head()

SK_ID_PREV  SK_ID_CURR NAME_CONTRACT_TYPE  AMT_ANNUITY  AMT_APPLICATION  \
0     2030495      271877     Consumer loans     1730.430          17145.0   
1     2802425      108129         Cash loans    25188.615         607500.0   
2     2523466      122040         Cash loans    15060.735         112500.0   
3     2819243      176158         Cash loans    47041.335         450000.0   
4     1784265      202054         Cash loans    31924.395         337500.0   

   AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE WEEKDAY_APPR_PROCESS_START  \
0     17145.0               0.0          17145.0                   SATURDAY   
1    679671.0               NaN         607500.0                   THURSDAY   
2    136444.5               NaN         112500.0                    TUESDAY   
3    470790.0               NaN         450000.0                     MONDAY   
4    404055.0               NaN         337500.0                   THURSDAY   

   HOUR_APPR_PROCESS_START  ... NAME_SELLER_INDUSTRY  CNT_PAYMENT  \
0                       15  ...         Connectivity         12.0   
1                       11  ...                  XNA         36.0   
2                       11  ...                  XNA         12.0   
3                        7  ...                  XNA         12.0   
4                        9  ...                  XNA         24.0   

   NAME_YIELD_GROUP       PRODUCT_COMBINATION  DAYS_FIRST_DRAWING  \
0            middle  POS mobile with interest            365243.0   
1        low_action          Cash X-Sell: low            365243.0   
2              high         Cash X-Sell: high            365243.0   
3            middle       Cash X-Sell: middle            365243.0   
4              high         Cash Street: high                 NaN   

  DAYS_FIRST_DUE DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE DAYS_TERMINATION  \
0          -42.0                     300.0          -42.0            -37.0   
1         -134.0                     916.0       365243.0         365243.0   
2         -271.0                      59.0       365243.0         365243.0   
3         -482.0                    -152.0         -182.0           -177.0   
4            NaN                       NaN            NaN              NaN   

  NFLAG_INSURED_ON_APPROVAL  
0                       0.0  
1                       1.0  
2                       1.0  
3                       1.0  
4                       NaN  

[5 rows x 37 columns]

In [ ]:
#Reading the application data 
df=pd.read_csv('/content/drive/MyDrive/Creditcard/application_data.csv')
df.shape

(307511, 122)

In [ ]:
#Loading it to the dataframe
applied_data=pd.read_csv('/content/drive/MyDrive/Creditcard/application_data.csv')

In [ ]:
#Displaying the first 5 rows and columns
applied_data.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                 0                0                0                0   
1  ...                 0                0                0                0   
2  ...                 0                0                0                0   
3  ...                 0                0                0                0   
4  ...                 0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        NaN                       NaN   
4                        0.0                       0.0   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         NaN                        NaN   
4                         0.0                        0.0   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         1.0  
1                        0.0                         0.0  
2                        0.0                         0.0  
3                        NaN                         NaN  
4                        0.0                         0.0  

[5 rows x 122 columns]

In [ ]:
applied_data.shape

(307511, 122)

The application data file consists of 307511 rows and 122 columns

In [ ]:
#Displays the information of the data provided like total number of rows and columns data types and the memory usage
applied_data.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Data columns (total 122 columns):
 #    Column                        Dtype  
---   ------                        -----  
 0    SK_ID_CURR                    int64  
 1    TARGET                        int64  
 2    NAME_CONTRACT_TYPE            object 
 3    CODE_GENDER                   object 
 4    FLAG_OWN_CAR                  object 
 5    FLAG_OWN_REALTY               object 
 6    CNT_CHILDREN                  int64  
 7    AMT_INCOME_TOTAL              float64
 8    AMT_CREDIT                    float64
 9    AMT_ANNUITY                   float64
 10   AMT_GOODS_PRICE               float64
 11   NAME_TYPE_SUITE               object 
 12   NAME_INCOME_TYPE              object 
 13   NAME_EDUCATION_TYPE           object 
 14   NAME_FAMILY_STATUS            object 
 15   NAME_HOUSING_TYPE             object 
 16   REGION_POPULATION_RELATIVE    float64
 17   DAYS_BIRTH                    int64  
 18   DA

In [ ]:
#the describe() method is a useful tool for quickly summarizing and understanding the basic characteristics of a dataset.
applied_data.describe()

SK_ID_CURR         TARGET   CNT_CHILDREN  AMT_INCOME_TOTAL  \
count  307511.000000  307511.000000  307511.000000      3.075110e+05   
mean   278180.518577       0.080729       0.417052      1.687979e+05   
std    102790.175348       0.272419       0.722121      2.371231e+05   
min    100002.000000       0.000000       0.000000      2.565000e+04   
25%    189145.500000       0.000000       0.000000      1.125000e+05   
50%    278202.000000       0.000000       0.000000      1.471500e+05   
75%    367142.500000       0.000000       1.000000      2.025000e+05   
max    456255.000000       1.000000      19.000000      1.170000e+08   

         AMT_CREDIT    AMT_ANNUITY  AMT_GOODS_PRICE  \
count  3.075110e+05  307499.000000     3.072330e+05   
mean   5.990260e+05   27108.573909     5.383962e+05   
std    4.024908e+05   14493.737315     3.694465e+05   
min    4.500000e+04    1615.500000     4.050000e+04   
25%    2.700000e+05   16524.000000     2.385000e+05   
50%    5.135310e+05   24903.000000     4.500000e+05   
75%    8.086500e+05   34596.000000     6.795000e+05   
max    4.050000e+06  258025.500000     4.050000e+06   

       REGION_POPULATION_RELATIVE     DAYS_BIRTH  DAYS_EMPLOYED  ...  \
count               307511.000000  307511.000000  307511.000000  ...   
mean                     0.020868  -16036.995067   63815.045904  ...   
std                      0.013831    4363.988632  141275.766519  ...   
min                      0.000290  -25229.000000  -17912.000000  ...   
25%                      0.010006  -19682.000000   -2760.000000  ...   
50%                      0.018850  -15750.000000   -1213.000000  ...   
75%                      0.028663  -12413.000000    -289.000000  ...   
max                      0.072508   -7489.000000  365243.000000  ...   

       FLAG_DOCUMENT_18  FLAG_DOCUMENT_19  FLAG_DOCUMENT_20  FLAG_DOCUMENT_21  \
count     307511.000000     307511.000000     307511.000000     307511.000000   
mean           0.008130          0.000595          0.000507          0.000335   
std            0.089798          0.024387          0.022518          0.018299   
min            0.000000          0.000000          0.000000          0.000000   
25%            0.000000          0.000000          0.000000          0.000000   
50%            0.000000          0.000000          0.000000          0.000000   
75%            0.000000          0.000000          0.000000          0.000000   
max            1.000000          1.000000          1.000000          1.000000   

       AMT_REQ_CREDIT_BUREAU_HOUR  AMT_REQ_CREDIT_BUREAU_DAY  \
count               265992.000000              265992.000000   
mean                     0.006402                   0.007000   
std                      0.083849                   0.110757   
min                      0.000000                   0.000000   
25%                      0.000000                   0.000000   
50%                      0.000000                   0.000000   
75%                      0.000000                   0.000000   
max                      4.000000                   9.000000   

       AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
count               265992.000000              265992.000000   
mean                     0.034362                   0.267395   
std                      0.204685                   0.916002   
min                      0.000000                   0.000000   
25%                      0.000000                   0.000000   
50%                      0.000000                   0.000000   
75%                      0.000000                   0.000000   
max                      8.000000                  27.000000   

       AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
count              265992.000000               265992.000000  
mean                    0.265474                    1.899974  
std                     0.794056                    1.869295  
min                     0.000000                    0.000000  
25%              

In [ ]:
#This method is used to check if a value in a Pandas DataFrame or Series object is null or missing.
applied_data.isnull().sum().head(50)

SK_ID_CURR                          0
TARGET                              0
NAME_CONTRACT_TYPE                  0
CODE_GENDER                         0
FLAG_OWN_CAR                        0
FLAG_OWN_REALTY                     0
CNT_CHILDREN                        0
AMT_INCOME_TOTAL                    0
AMT_CREDIT                          0
AMT_ANNUITY                        12
AMT_GOODS_PRICE                   278
NAME_TYPE_SUITE                  1292
NAME_INCOME_TYPE                    0
NAME_EDUCATION_TYPE                 0
NAME_FAMILY_STATUS                  0
NAME_HOUSING_TYPE                   0
REGION_POPULATION_RELATIVE          0
DAYS_BIRTH                          0
DAYS_EMPLOYED                       0
DAYS_REGISTRATION                   0
DAYS_ID_PUBLISH                     0
OWN_CAR_AGE                    202929
FLAG_MOBIL                          0
FLAG_EMP_PHONE                      0
FLAG_WORK_PHONE                     0
FLAG_CONT_MOBILE                    0
FLAG_PHONE  

In [ ]:
#Calucalting the percentage of missing values in specific rows
(applied_data.isnull().sum()/len(applied_data)*100).sort_values(ascending=False).head(50)

COMMONAREA_MEDI                 69.872297
COMMONAREA_AVG                  69.872297
COMMONAREA_MODE                 69.872297
NONLIVINGAPARTMENTS_MODE        69.432963
NONLIVINGAPARTMENTS_AVG         69.432963
NONLIVINGAPARTMENTS_MEDI        69.432963
FONDKAPREMONT_MODE              68.386172
LIVINGAPARTMENTS_MODE           68.354953
LIVINGAPARTMENTS_AVG            68.354953
LIVINGAPARTMENTS_MEDI           68.354953
FLOORSMIN_AVG                   67.848630
FLOORSMIN_MODE                  67.848630
FLOORSMIN_MEDI                  67.848630
YEARS_BUILD_MEDI                66.497784
YEARS_BUILD_MODE                66.497784
YEARS_BUILD_AVG                 66.497784
OWN_CAR_AGE                     65.990810
LANDAREA_MEDI                   59.376738
LANDAREA_MODE                   59.376738
LANDAREA_AVG                    59.376738
BASEMENTAREA_MEDI               58.515956
BASEMENTAREA_AVG                58.515956
BASEMENTAREA_MODE               58.515956
EXT_SOURCE_1                    56

In [ ]:
#Checking for the total number of columns which consists of missing values more than 35%
missing_values = applied_data.isna().mean()
mask = missing_values > 0.35
cols_with_high_missing = missing_values[mask]
num_cols_with_high_missing = cols_with_high_missing.shape[0]
print(f"There are {num_cols_with_high_missing} columns with more than 35% missing values.")

There are 49 columns with more than 35% missing values.


In [ ]:
null_col = applied_data.isnull().sum().sort_values(ascending = False)
null_col = null_col[null_col.values >(0.35*len(applied_data))]

In [ ]:
#Dropping those columns ad displays the total number of columns dropped
cols_to_drop = missing_values[mask].index.tolist()
applied_update = applied_data.drop(cols_to_drop, axis=1)
print(f"The updated data frame has {applied_update.shape[1]} columns after dropping columns with missing values greater than 35%")

The updated data frame has 73 columns after dropping columns with missing values greater than 35%


In [ ]:
#New dataframe after the columns are deleted
applied_update.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                 0                0                0                0   
1  ...                 0                0                0                0   
2  ...                 0                0                0                0   
3  ...                 0                0                0                0   
4  ...                 0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        NaN                       NaN   
4                        0.0                       0.0   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         NaN                        NaN   
4                         0.0                        0.0   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         1.0  
1                        0.0                         0.0  
2                        0.0                         0.0  
3                        NaN                         NaN  
4                        0.0                         0.0  

[5 rows x 73 columns]

In [ ]:
applied_update.shape

(307511, 73)

In [ ]:
#Checking the percentage of missing values present after dropping the columns
(applied_update.isnull().sum()/len(applied_update)*100).sort_values(ascending=False).head(50)

OCCUPATION_TYPE               31.345545
EXT_SOURCE_3                  19.825307
AMT_REQ_CREDIT_BUREAU_YEAR    13.501631
AMT_REQ_CREDIT_BUREAU_QRT     13.501631
AMT_REQ_CREDIT_BUREAU_MON     13.501631
AMT_REQ_CREDIT_BUREAU_WEEK    13.501631
AMT_REQ_CREDIT_BUREAU_DAY     13.501631
AMT_REQ_CREDIT_BUREAU_HOUR    13.501631
NAME_TYPE_SUITE                0.420148
OBS_30_CNT_SOCIAL_CIRCLE       0.332021
DEF_30_CNT_SOCIAL_CIRCLE       0.332021
OBS_60_CNT_SOCIAL_CIRCLE       0.332021
DEF_60_CNT_SOCIAL_CIRCLE       0.332021
EXT_SOURCE_2                   0.214626
AMT_GOODS_PRICE                0.090403
AMT_ANNUITY                    0.003902
CNT_FAM_MEMBERS                0.000650
DAYS_LAST_PHONE_CHANGE         0.000325
FLAG_DOCUMENT_17               0.000000
FLAG_DOCUMENT_18               0.000000
FLAG_DOCUMENT_21               0.000000
FLAG_DOCUMENT_20               0.000000
FLAG_DOCUMENT_19               0.000000
FLAG_DOCUMENT_2                0.000000
FLAG_DOCUMENT_3                0.000000


In [ ]:
# Check the data type of each column
applied_update.dtypes.head(50)

SK_ID_CURR                       int64
TARGET                           int64
NAME_CONTRACT_TYPE              object
CODE_GENDER                     object
FLAG_OWN_CAR                    object
FLAG_OWN_REALTY                 object
CNT_CHILDREN                     int64
AMT_INCOME_TOTAL               float64
AMT_CREDIT                     float64
AMT_ANNUITY                    float64
AMT_GOODS_PRICE                float64
NAME_TYPE_SUITE                 object
NAME_INCOME_TYPE                object
NAME_EDUCATION_TYPE             object
NAME_FAMILY_STATUS              object
NAME_HOUSING_TYPE               object
REGION_POPULATION_RELATIVE     float64
DAYS_BIRTH                       int64
DAYS_EMPLOYED                    int64
DAYS_REGISTRATION              float64
DAYS_ID_PUBLISH                  int64
FLAG_MOBIL                       int64
FLAG_EMP_PHONE                   int64
FLAG_WORK_PHONE                  int64
FLAG_CONT_MOBILE                 int64
FLAG_PHONE               

In [ ]:
#Imputing those missing values using the statistical methods
applied_update.AMT_REQ_CREDIT_BUREAU_YEAR.fillna(applied_update.AMT_REQ_CREDIT_BUREAU_YEAR.mode()[0],inplace = True)
applied_update.AMT_REQ_CREDIT_BUREAU_MON.fillna(applied_update.AMT_REQ_CREDIT_BUREAU_MON.mode()[0],inplace = True)   #AMT_REQ_CREDIT_BUREAU_MON  
applied_update.AMT_REQ_CREDIT_BUREAU_WEEK.fillna(applied_update.AMT_REQ_CREDIT_BUREAU_WEEK.mode()[0],inplace = True) #AMT_REQ_CREDIT_BUREAU_WEEK
applied_update.AMT_REQ_CREDIT_BUREAU_DAY.fillna(applied_update.AMT_REQ_CREDIT_BUREAU_DAY.mode()[0],inplace = True)   #AMT_REQ_CREDIT_BUREAU_DAY
applied_update.AMT_REQ_CREDIT_BUREAU_HOUR.fillna(applied_update.AMT_REQ_CREDIT_BUREAU_HOUR.mode()[0],inplace = True) #AMT_REQ_CREDIT_BUREAU_HOUR
applied_update.AMT_REQ_CREDIT_BUREAU_QRT.fillna(applied_update.AMT_REQ_CREDIT_BUREAU_QRT.mode()[0],inplace = True)   #AMT_REQ_CREDIT_BUREAU_QRT
applied_update.NAME_TYPE_SUITE.fillna(applied_update.NAME_TYPE_SUITE.mode()[0],inplace = True) #NAME_TYPE_SUITE
applied_update.OBS_30_CNT_SOCIAL_CIRCLE.fillna( applied_update.OBS_30_CNT_SOCIAL_CIRCLE.mode()[0],inplace = True) #OBS_30_CNT_SOCIAL_CIRCLE
applied_update.DEF_30_CNT_SOCIAL_CIRCLE.fillna( applied_update.DEF_30_CNT_SOCIAL_CIRCLE.mode()[0],inplace = True) #DEF_30_CNT_SOCIAL_CIRCLE
applied_update.OBS_60_CNT_SOCIAL_CIRCLE.fillna( applied_update.OBS_60_CNT_SOCIAL_CIRCLE.mode()[0],inplace = True) #OBS_60_CNT_SOCIAL_CIRCLE
applied_update.DEF_60_CNT_SOCIAL_CIRCLE.fillna( applied_update.DEF_60_CNT_SOCIAL_CIRCLE.mode()[0],inplace = True) #DEF_60_CNT_SOCIAL_CIRCLE
applied_update.CNT_FAM_MEMBERS.fillna(applied_update.CNT_FAM_MEMBERS.mode() , inplace = True) #CNT_FAM_MEMBERS
applied_update.DAYS_LAST_PHONE_CHANGE.fillna(applied_update.DAYS_LAST_PHONE_CHANGE.mode()[0],inplace = True) #DAYS_LAST_PHONE_CHANGE
applied_update.EXT_SOURCE_2.fillna(applied_update.EXT_SOURCE_2.median() , inplace = True) #EXT_SOURCE_2
applied_update.EXT_SOURCE_3.fillna(applied_update.EXT_SOURCE_3.median() , inplace = True) #EXT_SOURCE_3


In [ ]:
#Checking for the percentage of the missing values after imputing the missing values
(applied_update.isnull().sum()/len(applied_update)*100).sort_values(ascending=False).head(50)

OCCUPATION_TYPE               31.345545
AMT_GOODS_PRICE                0.090403
AMT_ANNUITY                    0.003902
CNT_FAM_MEMBERS                0.000650
FLAG_DOCUMENT_3                0.000000
FLAG_DOCUMENT_8                0.000000
FLAG_DOCUMENT_7                0.000000
FLAG_DOCUMENT_6                0.000000
FLAG_DOCUMENT_5                0.000000
FLAG_DOCUMENT_4                0.000000
SK_ID_CURR                     0.000000
FLAG_DOCUMENT_10               0.000000
FLAG_DOCUMENT_2                0.000000
DAYS_LAST_PHONE_CHANGE         0.000000
DEF_60_CNT_SOCIAL_CIRCLE       0.000000
OBS_60_CNT_SOCIAL_CIRCLE       0.000000
DEF_30_CNT_SOCIAL_CIRCLE       0.000000
OBS_30_CNT_SOCIAL_CIRCLE       0.000000
EXT_SOURCE_3                   0.000000
FLAG_DOCUMENT_9                0.000000
FLAG_DOCUMENT_11               0.000000
ORGANIZATION_TYPE              0.000000
FLAG_DOCUMENT_20               0.000000
AMT_REQ_CREDIT_BUREAU_QRT      0.000000
AMT_REQ_CREDIT_BUREAU_MON      0.000000


In [ ]:
# Check if the dataset contains any negative values in the numerical columns
numerical_cols = applied_update.select_dtypes(include=[np.number]).columns
if (applied_update[numerical_cols] < 0).any().any():
    print("The dataset contains negative values in the numerical columns.")
else:
    print("The dataset does not contain negative values in the numerical columns.")

The dataset contains negative values in the numerical columns.


In [ ]:
# Check if the dataset contains any negative values
def contains_negative_values(column):
    if column.dtype.kind in 'biufc':
        return (column < 0).any()
    else:
        return False

columns_with_negative_values = []
for column_name, column in applied_update.iteritems():
    if contains_negative_values(column):
        columns_with_negative_values.append(column_name)

if len(columns_with_negative_values) > 0:
    print(f"The dataset contains negative values in the following columns: {columns_with_negative_values}")
else:
    print("The dataset does not contain negative values.")

The dataset contains negative values in the following columns: ['DAYS_BIRTH', 'DAYS_EMPLOYED', 'DAYS_REGISTRATION', 'DAYS_ID_PUBLISH', 'DAYS_LAST_PHONE_CHANGE']


In the code above, we define a function contains_negative_values that takes a pandas Series as input and returns True if the series contains negative values, and False otherwise. The function checks if the data type of the series is a numeric data type ('b' for boolean, 'i' for integer, 'u' for unsigned integer, 'f' for float, and 'c' for complex), and if it is, it checks if any element in the series is negative.

In [ ]:
#Overview of the data that contais the negative values
applied_update[['DAYS_BIRTH','DAYS_EMPLOYED','DAYS_REGISTRATION','DAYS_ID_PUBLISH','DAYS_LAST_PHONE_CHANGE']]

DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  DAYS_ID_PUBLISH  \
0            -9461           -637            -3648.0            -2120   
1           -16765          -1188            -1186.0             -291   
2           -19046           -225            -4260.0            -2531   
3           -19005          -3039            -9833.0            -2437   
4           -19932          -3038            -4311.0            -3458   
...            ...            ...                ...              ...   
307506       -9327           -236            -8456.0            -1982   
307507      -20775         365243            -4388.0            -4090   
307508      -14966          -7921            -6737.0            -5150   
307509      -11961          -4786            -2562.0             -931   
307510      -16856          -1262            -5128.0             -410   

        DAYS_LAST_PHONE_CHANGE  
0                      -1134.0  
1                       -828.0  
2                       -815.0  
3                       -617.0  
4                      -1106.0  
...                        ...  
307506                  -273.0  
307507                     0.0  
307508                 -1909.0  
307509                  -322.0  
307510                  -787.0  

[307511 rows x 5 columns]

In [ ]:
# Convert selected columns with negative values to positive values
selected_columns = ['DAYS_BIRTH','DAYS_EMPLOYED','DAYS_REGISTRATION','DAYS_ID_PUBLISH','DAYS_LAST_PHONE_CHANGE']
applied_update[selected_columns] = applied_update[selected_columns].abs()

In [ ]:
# Check if the dataset contains any negative values in the numerical columns
numerical_cols = applied_update.select_dtypes(include=[np.number]).columns
if (applied_update[numerical_cols] < 0).any().any():
    print("The dataset contains negative values in the numerical columns.")
else:
    print("The dataset does not contain negative values in the numerical columns.")

The dataset does not contain negative values in the numerical columns.


In [ ]:
#Overview of the columns after treating the negative values
applied_update[['DAYS_BIRTH','DAYS_EMPLOYED','DAYS_REGISTRATION','DAYS_ID_PUBLISH','DAYS_LAST_PHONE_CHANGE']]

DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  DAYS_ID_PUBLISH  \
0             9461            637             3648.0             2120   
1            16765           1188             1186.0              291   
2            19046            225             4260.0             2531   
3            19005           3039             9833.0             2437   
4            19932           3038             4311.0             3458   
...            ...            ...                ...              ...   
307506        9327            236             8456.0             1982   
307507       20775         365243             4388.0             4090   
307508       14966           7921             6737.0             5150   
307509       11961           4786             2562.0              931   
307510       16856           1262             5128.0              410   

        DAYS_LAST_PHONE_CHANGE  
0                       1134.0  
1                        828.0  
2                        815.0  
3                        617.0  
4                       1106.0  
...                        ...  
307506                   273.0  
307507                     0.0  
307508                  1909.0  
307509                   322.0  
307510                   787.0  

[307511 rows x 5 columns]

In [ ]:
#Changing the values to 0s and 1s from the columns 
applied_update['FLAG_OWN_CAR'] = np.where(applied_update['FLAG_OWN_CAR']=='Y', 1 , 0)  
applied_update['FLAG_OWN_REALTY'] = np.where(applied_update['FLAG_OWN_REALTY']=='Y', 1 , 0)

In [ ]:
applied_update[['FLAG_OWN_CAR','FLAG_OWN_REALTY']].head()

FLAG_OWN_CAR  FLAG_OWN_REALTY
0             0                1
1             0                0
2             1                1
3             0                1
4             0                1

In [ ]:
#Displays the total number of people working in different organizations
applied_update.ORGANIZATION_TYPE.value_counts().head()

Business Entity Type 3    67992
XNA                       55374
Self-employed             38412
Other                     16683
Medicine                  11193
Name: ORGANIZATION_TYPE, dtype: int64

In [ ]:
applied_update.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER  FLAG_OWN_CAR  \
0      100002       1         Cash loans           M             0   
1      100003       0         Cash loans           F             0   
2      100004       0    Revolving loans           M             1   
3      100006       0         Cash loans           F             0   
4      100007       0         Cash loans           M             0   

   FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0                1             0          202500.0    406597.5      24700.5   
1                0             0          270000.0   1293502.5      35698.5   
2                1             0           67500.0    135000.0       6750.0   
3                1             0          135000.0    312682.5      29686.5   
4                1             0          121500.0    513000.0      21865.5   

   ...  FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21  \
0  ...                 0                0                0                0   
1  ...                 0                0                0                0   
2  ...                 0                0                0                0   
3  ...                 0                0                0                0   
4  ...                 0                0                0                0   

  AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                        0.0                       0.0   
1                        0.0                       0.0   
2                        0.0                       0.0   
3                        0.0                       0.0   
4                        0.0                       0.0   

   AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                         0.0                        0.0   
1                         0.0                        0.0   
2                         0.0                        0.0   
3                         0.0                        0.0   
4                         0.0                        0.0   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  
0                        0.0                         1.0  
1                        0.0                         0.0  
2                        0.0                         0.0  
3                        0.0                         0.0  
4                        0.0                         0.0  

[5 rows x 73 columns]

In [ ]:
#Displays the total number of gender counts 
applied_update.CODE_GENDER.value_counts()

F      202448
M      105059
XNA         4
Name: CODE_GENDER, dtype: int64

In [ ]:
#From the above count we can see that it consists of XNA values which may be null so changing it to the values with more number of counts
applied_update.loc[applied_update.CODE_GENDER == 'XNA','CODE_GENDER'] = 'F' 
applied_update.CODE_GENDER.value_counts()

F    202452
M    105059
Name: CODE_GENDER, dtype: int64

In [ ]:
applied_update.ORGANIZATION_TYPE.value_counts().head()

Business Entity Type 3    67992
XNA                       55374
Self-employed             38412
Other                     16683
Medicine                  11193
Name: ORGANIZATION_TYPE, dtype: int64

In [ ]:
#Displays the total number of people working in differet organaization and thier income types
applied_update[['ORGANIZATION_TYPE','NAME_INCOME_TYPE']].head(30)

ORGANIZATION_TYPE      NAME_INCOME_TYPE
0   Business Entity Type 3               Working
1                   School         State servant
2               Government               Working
3   Business Entity Type 3               Working
4                 Religion               Working
5                    Other         State servant
6   Business Entity Type 3  Commercial associate
7                    Other         State servant
8                      XNA             Pensioner
9              Electricity               Working
10                Medicine               Working
11                     XNA             Pensioner
12  Business Entity Type 2               Working
13           Self-employed               Working
14       Transport: type 2               Working
15  Business Entity Type 2               Working
16              Government               Working
17            Construction               Working
18                 Housing               Working
19            Kindergarten         State servant
20           Self-employed               Working
21           Trade: type 7  Commercial associate
22           Self-employed               Working
23                     XNA             Pensioner
24  Business Entity Type 3               Working
25  Business Entity Type 3               Working
26  Business Entity Type 3               Working
27       Industry: type 11               Working
28                Military         State servant
29  Business Entity Type 3               Working

In [ ]:
applied_update.NAME_INCOME_TYPE.value_counts()    #Check the counts for each in NAME_INCOME_TYPE

Working                 158774
Commercial associate     71617
Pensioner                55362
State servant            21703
Unemployed                  22
Student                     18
Businessman                 10
Maternity leave              5
Name: NAME_INCOME_TYPE, dtype: int64

In [ ]:
#Some values in organizattion types shows XNA which may be null values replacing those with the pensioner
applied_update['ORGANIZATION_TYPE'] = applied_update['ORGANIZATION_TYPE'].replace('XNA', 'Pensioner')
applied_update['OCCUPATION_TYPE'].fillna('Pensioner' , inplace = True)

In [ ]:
#Displaying the count after replacing
applied_update.ORGANIZATION_TYPE.value_counts().head()

Business Entity Type 3    67992
Pensioner                 55374
Self-employed             38412
Other                     16683
Medicine                  11193
Name: ORGANIZATION_TYPE, dtype: int64

In [ ]:
# Converting the age to years and the datatype to int
applied_update.DAYS_BIRTH = ((applied_update.DAYS_BIRTH)/365).astype('int')

In [ ]:
applied_update.DAYS_BIRTH.describe() 

count    307511.000000
mean         43.435968
std          11.954593
min          20.000000
25%          34.000000
50%          43.000000
75%          53.000000
max          69.000000
Name: DAYS_BIRTH, dtype: float64

We can see that the distribution of age among applicants is as follows :
Minimum age applicant is 20 years and max is 69 years.

In [ ]:
#We can bin the DAYS_BIRTH column to get the different buckets of age that have applied for the loan.
applied_update['DAYS_BIRTH_BINS']=pd.cut(applied_update.DAYS_BIRTH, bins=[19,40,60,100], labels=['Young_Age','Middle_Age','Senior_Citizen'])

In [ ]:
applied_update.DAYS_BIRTH_BINS.value_counts()

Middle_Age        142220
Young_Age         135923
Senior_Citizen     29368
Name: DAYS_BIRTH_BINS, dtype: int64

In [ ]:
# checking statistics for AMT_INCOME_TOTAL column
applied_update.AMT_INCOME_TOTAL.describe()

count    3.075110e+05
mean     1.687979e+05
std      2.371231e+05
min      2.565000e+04
25%      1.125000e+05
50%      1.471500e+05
75%      2.025000e+05
max      1.170000e+08
Name: AMT_INCOME_TOTAL, dtype: float64

In [ ]:
applied_update.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Data columns (total 74 columns):
 #   Column                       Non-Null Count   Dtype   
---  ------                       --------------   -----   
 0   SK_ID_CURR                   307511 non-null  int64   
 1   TARGET                       307511 non-null  int64   
 2   NAME_CONTRACT_TYPE           307511 non-null  object  
 3   CODE_GENDER                  307511 non-null  object  
 4   FLAG_OWN_CAR                 307511 non-null  int64   
 5   FLAG_OWN_REALTY              307511 non-null  int64   
 6   CNT_CHILDREN                 307511 non-null  int64   
 7   AMT_INCOME_TOTAL             307511 non-null  float64 
 8   AMT_CREDIT                   307511 non-null  float64 
 9   AMT_ANNUITY                  307499 non-null  float64 
 10  AMT_GOODS_PRICE              307233 non-null  float64 
 11  NAME_TYPE_SUITE              307511 non-null  object  
 12  NAME_INCOME_TYPE             307511 non-null

In [ ]:
applied_update['NAME_CONTRACT_TYPE'] = applied_update['NAME_CONTRACT_TYPE'].astype('category')
applied_update['CODE_GENDER'] = applied_update['CODE_GENDER'].astype('category')
applied_update['NAME_TYPE_SUITE'] = applied_update['NAME_TYPE_SUITE'].astype('category')
applied_update['NAME_INCOME_TYPE'] = applied_update['NAME_INCOME_TYPE'].astype('category')
applied_update['NAME_EDUCATION_TYPE'] = applied_update['NAME_EDUCATION_TYPE'].astype('category')
applied_update['NAME_FAMILY_STATUS'] = applied_update['NAME_FAMILY_STATUS'].astype('category')
applied_update['NAME_HOUSING_TYPE'] = applied_update['NAME_HOUSING_TYPE'].astype('category')
applied_update['OCCUPATION_TYPE'] = applied_update['OCCUPATION_TYPE'].astype('category')
applied_update['WEEKDAY_APPR_PROCESS_START'] = applied_update['WEEKDAY_APPR_PROCESS_START'].astype('category')
applied_update['ORGANIZATION_TYPE'] = applied_update['ORGANIZATION_TYPE'].astype('category')

In [ ]:
#deleting some coulmns which are not useful for the further process
unwanted=['FLAG_MOBIL', 'FLAG_EMP_PHONE', 'FLAG_WORK_PHONE', 'FLAG_CONT_MOBILE',
       'FLAG_PHONE', 'FLAG_EMAIL','REGION_RATING_CLIENT','REGION_RATING_CLIENT_W_CITY','FLAG_EMAIL', 'REGION_RATING_CLIENT',
       'REGION_RATING_CLIENT_W_CITY', 'FLAG_DOCUMENT_2', 'FLAG_DOCUMENT_3','FLAG_DOCUMENT_4', 'FLAG_DOCUMENT_5', 'FLAG_DOCUMENT_6',
       'FLAG_DOCUMENT_7', 'FLAG_DOCUMENT_8', 'FLAG_DOCUMENT_9','FLAG_DOCUMENT_10', 'FLAG_DOCUMENT_11', 'FLAG_DOCUMENT_12',
       'FLAG_DOCUMENT_13', 'FLAG_DOCUMENT_14', 'FLAG_DOCUMENT_15','FLAG_DOCUMENT_16', 'FLAG_DOCUMENT_17', 'FLAG_DOCUMENT_18',
       'FLAG_DOCUMENT_19', 'FLAG_DOCUMENT_20', 'FLAG_DOCUMENT_21']

applied_update.drop(labels=unwanted,axis=1,inplace=True)

In [ ]:
applied_update.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER  FLAG_OWN_CAR  \
0      100002       1         Cash loans           M             0   
1      100003       0         Cash loans           F             0   
2      100004       0    Revolving loans           M             1   
3      100006       0         Cash loans           F             0   
4      100007       0         Cash loans           M             0   

   FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0                1             0          202500.0    406597.5      24700.5   
1                0             0          270000.0   1293502.5      35698.5   
2                1             0           67500.0    135000.0       6750.0   
3                1             0          135000.0    312682.5      29686.5   
4                1             0          121500.0    513000.0      21865.5   

   ...  OBS_60_CNT_SOCIAL_CIRCLE DEF_60_CNT_SOCIAL_CIRCLE  \
0  ...                       2.0                      2.0   
1  ...                       1.0                      0.0   
2  ...                       0.0                      0.0   
3  ...                       2.0                      0.0   
4  ...                       0.0                      0.0   

  DAYS_LAST_PHONE_CHANGE AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY  \
0                 1134.0                        0.0                       0.0   
1                  828.0                        0.0                       0.0   
2                  815.0                        0.0                       0.0   
3                  617.0                        0.0                       0.0   
4                 1106.0                        0.0                       0.0   

  AMT_REQ_CREDIT_BUREAU_WEEK  AMT_REQ_CREDIT_BUREAU_MON  \
0                        0.0                        0.0   
1                        0.0                        0.0   
2                        0.0                        0.0   
3                        0.0                        0.0   
4                        0.0                        0.0   

   AMT_REQ_CREDIT_BUREAU_QRT  AMT_REQ_CREDIT_BUREAU_YEAR  DAYS_BIRTH_BINS  
0                        0.0                         1.0        Young_Age  
1                        0.0                         0.0       Middle_Age  
2                        0.0                         0.0       Middle_Age  
3                        0.0                         0.0       Middle_Age  
4                        0.0                         0.0       Middle_Age  

[5 rows x 46 columns]

In [ ]:
#Total number of numerical coulumns after deeting the unwanted columns from the dataset
numerical_col = applied_update.select_dtypes(include='number').columns
len(numerical_col)

35

In [ ]:
#checking fpr the target value counts
#1 represents the people with payment difficulties
#0 reprents the people with the other cases
applied_data['TARGET'].value_counts()

0    282686
1     24825
Name: TARGET, dtype: int64

In [ ]:
#Naming the people with payment difficulties as defaulters and 0 with non defaulsters
defaulters=applied_update[applied_update.TARGET==1]
nondefaulters=applied_update[applied_update.TARGET==0]

In [ ]:
applied_update['DAYS_EMPLOYED'].describe()

count    307511.000000
mean      67724.742149
std      139443.751806
min           0.000000
25%         933.000000
50%        2219.000000
75%        5707.000000
max      365243.000000
Name: DAYS_EMPLOYED, dtype: float64

There is an outlier here. The max value is 365243 days which is not practically possible. This might be an error and we can replace this value with null

In [ ]:
applied_update['DAYS_EMPLOYED']=applied_update['DAYS_EMPLOYED'].replace(365243, np.nan)
applied_update['DAYS_EMPLOYED'].describe()

count    252137.000000
mean       2384.169325
std        2338.360162
min           0.000000
25%         767.000000
50%        1648.000000
75%        3175.000000
max       17912.000000
Name: DAYS_EMPLOYED, dtype: float64

In [ ]:
#Deriving variable "Years Employed" from days employed
applied_update['YEARS_EMPLOYED'] = (applied_update['DAYS_EMPLOYED']/365)

#update dataframes
defaulters=applied_update[applied_update.TARGET==1] 
nondefaulters=applied_update[applied_update.TARGET==0]

In [ ]:
#selecting columns for correlation, removing cols for floor and house ec

cols=[ 'EXT_SOURCE_3', 'EXT_SOURCE_2',
       'AMT_GOODS_PRICE', 'AMT_ANNUITY', 'CNT_FAM_MEMBERS',
       'DAYS_LAST_PHONE_CHANGE', 'AMT_CREDIT', 'AMT_INCOME_TOTAL',
       'DAYS_REGISTRATION', 'REGION_POPULATION_RELATIVE','CNT_CHILDREN', 'HOUR_APPR_PROCESS_START',
      
       'DAYS_ID_PUBLISH', 'DAYS_EMPLOYED', 'DAYS_BIRTH']

In [ ]:
defaulters_1=defaulters[cols]
defaulters_correlation = defaulters_1.corr()
round(defaulters_correlation, 3)

EXT_SOURCE_3  EXT_SOURCE_2  AMT_GOODS_PRICE  \
EXT_SOURCE_3                       1.000         0.045            0.049   
EXT_SOURCE_2                       0.045         1.000            0.130   
AMT_GOODS_PRICE                    0.049         0.130            1.000   
AMT_ANNUITY                        0.031         0.115            0.753   
CNT_FAM_MEMBERS                   -0.031         0.002            0.047   
DAYS_LAST_PHONE_CHANGE             0.025         0.207            0.118   
AMT_CREDIT                         0.048         0.120            0.983   
AMT_INCOME_TOTAL                  -0.019         0.007            0.038   
DAYS_REGISTRATION                  0.070         0.070            0.026   
REGION_POPULATION_RELATIVE        -0.010         0.170            0.076   
CNT_CHILDREN                      -0.025        -0.012           -0.008   
HOUR_APPR_PROCESS_START           -0.054         0.134            0.044   
DAYS_ID_PUBLISH                    0.080         0.059            0.056   
DAYS_EMPLOYED                      0.050         0.094            0.112   
DAYS_BIRTH                         0.130         0.108            0.136   

                            AMT_ANNUITY  CNT_FAM_MEMBERS  \
EXT_SOURCE_3                      0.031           -0.031   
EXT_SOURCE_2                      0.115            0.002   
AMT_GOODS_PRICE                   0.753            0.047   
AMT_ANNUITY                       1.000            0.076   
CNT_FAM_MEMBERS                   0.076            1.000   
DAYS_LAST_PHONE_CHANGE            0.080            0.013   
AMT_CREDIT                        0.752            0.051   
AMT_INCOME_TOTAL                  0.046            0.007   
DAYS_REGISTRATION                -0.034           -0.146   
REGION_POPULATION_RELATIVE        0.072           -0.030   
CNT_CHILDREN                      0.031            0.885   
HOUR_APPR_PROCESS_START           0.031           -0.028   
DAYS_ID_PUBLISH                   0.017            0.032   
DAYS_EMPLOYED                     0.049            0.002   
DAYS_BIRTH                        0.014           -0.203   

                            DAYS_LAST_PHONE_CHANGE  AMT_CREDIT  \
EXT_SOURCE_3                                 0.025       0.048   
EXT_SOURCE_2                                 0.207       0.120   
AMT_GOODS_PRICE                              0.118       0.983   
AMT_ANNUITY                                  0.080       0.752   
CNT_FAM_MEMBERS                              0.013       0.051   
DAYS_LAST_PHONE_CHANGE                       1.000       0.111   
AMT_CREDIT                                   0.111       1.000   
AMT_INCOME_TOTAL                             0.002       0.038   
DAYS_REGISTRATION                            0.072       0.026   
REGION_POPULATION_RELATIVE                   0.055       0.069   
CNT_CHILDREN                                -0.012      -0.002   
HOUR_APPR_PROCESS_START                      0.023       0.032   
DAYS_ID_PUBLISH                              0.124       0.052   
DAYS_EMPLOYED                                0.170       0.106   
DAYS_BIRTH                                   0.111       0.135   

                            AMT_INCOME_TOTAL  DAYS_REGISTRATION  \
EXT_SOURCE_3                          -0.019              0.070   
EXT_SOURCE_2                           0.007              0.070   
AMT_GOODS_PRICE                        0.038              0.026   
AMT_ANNUITY                            0.046             -0.034   
CNT_FAM_MEMBERS                        0.007             -0.146   
DAYS_LAST_PHONE_CHANGE                 0.002              0.072   
AMT_CREDIT                             0.038              0.026   
AMT_INCOME_TOTAL                       1.000             -0.000   
DAYS_REGISTRATION                     -0.000              1.000   
REGION_POPULATION_RELATIVE             0.009              0.056   
CNT_CHILDREN                           0.005             -0.149   
HOUR_



### Analysis of Previous Application Dataset



In [ ]:
#importing data from CSV file into pandas dataframe

import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt

 
previous_data = pd.read_csv('/content/drive/MyDrive/Creditcard/previous_application.csv')
previous_data.head()

SK_ID_PREV  SK_ID_CURR NAME_CONTRACT_TYPE  AMT_ANNUITY  AMT_APPLICATION  \
0     2030495      271877     Consumer loans     1730.430          17145.0   
1     2802425      108129         Cash loans    25188.615         607500.0   
2     2523466      122040         Cash loans    15060.735         112500.0   
3     2819243      176158         Cash loans    47041.335         450000.0   
4     1784265      202054         Cash loans    31924.395         337500.0   

   AMT_CREDIT  AMT_DOWN_PAYMENT  AMT_GOODS_PRICE WEEKDAY_APPR_PROCESS_START  \
0     17145.0               0.0          17145.0                   SATURDAY   
1    679671.0               NaN         607500.0                   THURSDAY   
2    136444.5               NaN         112500.0                    TUESDAY   
3    470790.0               NaN         450000.0                     MONDAY   
4    404055.0               NaN         337500.0                   THURSDAY   

   HOUR_APPR_PROCESS_START  ... NAME_SELLER_INDUSTRY  CNT_PAYMENT  \
0                       15  ...         Connectivity         12.0   
1                       11  ...                  XNA         36.0   
2                       11  ...                  XNA         12.0   
3                        7  ...                  XNA         12.0   
4                        9  ...                  XNA         24.0   

   NAME_YIELD_GROUP       PRODUCT_COMBINATION  DAYS_FIRST_DRAWING  \
0            middle  POS mobile with interest            365243.0   
1        low_action          Cash X-Sell: low            365243.0   
2              high         Cash X-Sell: high            365243.0   
3            middle       Cash X-Sell: middle            365243.0   
4              high         Cash Street: high                 NaN   

  DAYS_FIRST_DUE DAYS_LAST_DUE_1ST_VERSION  DAYS_LAST_DUE DAYS_TERMINATION  \
0          -42.0                     300.0          -42.0            -37.0   
1         -134.0                     916.0       365243.0         365243.0   
2         -271.0                      59.0       365243.0         365243.0   
3         -482.0                    -152.0         -182.0           -177.0   
4            NaN                       NaN            NaN              NaN   

  NFLAG_INSURED_ON_APPROVAL  
0                       0.0  
1                       1.0  
2                       1.0  
3                       1.0  
4                       NaN  

[5 rows x 37 columns]

In [ ]:
previous_data.shape

(1670214, 37)

In [ ]:
previous_data.dtypes.value_counts()

object     16
float64    15
int64       6
dtype: int64

In [ ]:
previous_data.columns

Index(['SK_ID_PREV', 'SK_ID_CURR', 'NAME_CONTRACT_TYPE', 'AMT_ANNUITY',
       'AMT_APPLICATION', 'AMT_CREDIT', 'AMT_DOWN_PAYMENT', 'AMT_GOODS_PRICE',
       'WEEKDAY_APPR_PROCESS_START', 'HOUR_APPR_PROCESS_START',
       'FLAG_LAST_APPL_PER_CONTRACT', 'NFLAG_LAST_APPL_IN_DAY',
       'RATE_DOWN_PAYMENT', 'RATE_INTEREST_PRIMARY',
       'RATE_INTEREST_PRIVILEGED', 'NAME_CASH_LOAN_PURPOSE',
       'NAME_CONTRACT_STATUS', 'DAYS_DECISION', 'NAME_PAYMENT_TYPE',
       'CODE_REJECT_REASON', 'NAME_TYPE_SUITE', 'NAME_CLIENT_TYPE',
       'NAME_GOODS_CATEGORY', 'NAME_PORTFOLIO', 'NAME_PRODUCT_TYPE',
       'CHANNEL_TYPE', 'SELLERPLACE_AREA', 'NAME_SELLER_INDUSTRY',
       'CNT_PAYMENT', 'NAME_YIELD_GROUP', 'PRODUCT_COMBINATION',
       'DAYS_FIRST_DRAWING', 'DAYS_FIRST_DUE', 'DAYS_LAST_DUE_1ST_VERSION',
       'DAYS_LAST_DUE', 'DAYS_TERMINATION', 'NFLAG_INSURED_ON_APPROVAL'],
      dtype='object')

In [ ]:
previous_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1670214 entries, 0 to 1670213
Data columns (total 37 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   SK_ID_PREV                   1670214 non-null  int64  
 1   SK_ID_CURR                   1670214 non-null  int64  
 2   NAME_CONTRACT_TYPE           1670214 non-null  object 
 3   AMT_ANNUITY                  1297979 non-null  float64
 4   AMT_APPLICATION              1670214 non-null  float64
 5   AMT_CREDIT                   1670213 non-null  float64
 6   AMT_DOWN_PAYMENT             774370 non-null   float64
 7   AMT_GOODS_PRICE              1284699 non-null  float64
 8   WEEKDAY_APPR_PROCESS_START   1670214 non-null  object 
 9   HOUR_APPR_PROCESS_START      1670214 non-null  int64  
 10  FLAG_LAST_APPL_PER_CONTRACT  1670214 non-null  object 
 11  NFLAG_LAST_APPL_IN_DAY       1670214 non-null  int64  
 12  RATE_DOWN_PAYMENT            774370 non-nu

Identifying missing values and filtering out columns with high missing values

In [ ]:
# checking columns with missing values
previous_data.isnull().sum()

SK_ID_PREV                           0
SK_ID_CURR                           0
NAME_CONTRACT_TYPE                   0
AMT_ANNUITY                     372235
AMT_APPLICATION                      0
AMT_CREDIT                           1
AMT_DOWN_PAYMENT                895844
AMT_GOODS_PRICE                 385515
WEEKDAY_APPR_PROCESS_START           0
HOUR_APPR_PROCESS_START              0
FLAG_LAST_APPL_PER_CONTRACT          0
NFLAG_LAST_APPL_IN_DAY               0
RATE_DOWN_PAYMENT               895844
RATE_INTEREST_PRIMARY          1664263
RATE_INTEREST_PRIVILEGED       1664263
NAME_CASH_LOAN_PURPOSE               0
NAME_CONTRACT_STATUS                 0
DAYS_DECISION                        0
NAME_PAYMENT_TYPE                    0
CODE_REJECT_REASON                   0
NAME_TYPE_SUITE                 820405
NAME_CLIENT_TYPE                     0
NAME_GOODS_CATEGORY                  0
NAME_PORTFOLIO                       0
NAME_PRODUCT_TYPE                    0
CHANNEL_TYPE             

In [ ]:
(previous_data.isnull().sum()/len(previous_data)*100).sort_values(ascending=False).head(50)

RATE_INTEREST_PRIVILEGED       99.643698
RATE_INTEREST_PRIMARY          99.643698
AMT_DOWN_PAYMENT               53.636480
RATE_DOWN_PAYMENT              53.636480
NAME_TYPE_SUITE                49.119754
NFLAG_INSURED_ON_APPROVAL      40.298129
DAYS_TERMINATION               40.298129
DAYS_LAST_DUE                  40.298129
DAYS_LAST_DUE_1ST_VERSION      40.298129
DAYS_FIRST_DUE                 40.298129
DAYS_FIRST_DRAWING             40.298129
AMT_GOODS_PRICE                23.081773
AMT_ANNUITY                    22.286665
CNT_PAYMENT                    22.286366
PRODUCT_COMBINATION             0.020716
AMT_CREDIT                      0.000060
NAME_YIELD_GROUP                0.000000
NAME_PORTFOLIO                  0.000000
NAME_SELLER_INDUSTRY            0.000000
SELLERPLACE_AREA                0.000000
CHANNEL_TYPE                    0.000000
NAME_PRODUCT_TYPE               0.000000
SK_ID_PREV                      0.000000
NAME_GOODS_CATEGORY             0.000000
NAME_CLIENT_TYPE

In [ ]:
# dropping unnecessary columns

previous_data.drop(['DAYS_FIRST_DRAWING','DAYS_FIRST_DUE','DAYS_LAST_DUE_1ST_VERSION','DAYS_LAST_DUE','DAYS_TERMINATION',
'NFLAG_INSURED_ON_APPROVAL','CNT_PAYMENT','NAME_TYPE_SUITE','RATE_INTEREST_PRIMARY','RATE_INTEREST_PRIVILEGED',
'RATE_DOWN_PAYMENT','AMT_GOODS_PRICE'],axis=1,inplace=True)

In [ ]:
(previous_data.isnull().sum()/len(previous_data)*100).sort_values(ascending=False).head(50)

AMT_DOWN_PAYMENT               53.636480
AMT_ANNUITY                    22.286665
PRODUCT_COMBINATION             0.020716
AMT_CREDIT                      0.000060
NAME_PAYMENT_TYPE               0.000000
NAME_YIELD_GROUP                0.000000
NAME_SELLER_INDUSTRY            0.000000
SELLERPLACE_AREA                0.000000
CHANNEL_TYPE                    0.000000
NAME_PRODUCT_TYPE               0.000000
NAME_PORTFOLIO                  0.000000
NAME_GOODS_CATEGORY             0.000000
NAME_CLIENT_TYPE                0.000000
CODE_REJECT_REASON              0.000000
SK_ID_PREV                      0.000000
DAYS_DECISION                   0.000000
SK_ID_CURR                      0.000000
NAME_CASH_LOAN_PURPOSE          0.000000
NFLAG_LAST_APPL_IN_DAY          0.000000
FLAG_LAST_APPL_PER_CONTRACT     0.000000
HOUR_APPR_PROCESS_START         0.000000
WEEKDAY_APPR_PROCESS_START      0.000000
AMT_APPLICATION                 0.000000
NAME_CONTRACT_TYPE              0.000000
NAME_CONTRACT_ST

In [ ]:
len(previous_data[(previous_data.AMT_APPLICATION==0.0) & (previous_data.AMT_ANNUITY.isnull())]) 
# these many columns are null for AMT_ANNUITY which have AMT_APPLICATION as 0.0 value
# so replacing these null values for AMT_ANNUITY with 0.0

335146

In [ ]:
# Replacing missing values for AMT_ANNUITY with 0.0 
previous_data['AMT_ANNUITY']=previous_data.AMT_ANNUITY.fillna(0.0)

In [ ]:
# Finding the mode of AMT_DOWN_PAYMENT to replace the missing values
previous_data.AMT_DOWN_PAYMENT.mode()[0]

0.0

In [ ]:
# Filling the missing values with mode
previous_data['AMT_DOWN_PAYMENT']=previous_data.AMT_DOWN_PAYMENT.fillna(previous_data.AMT_DOWN_PAYMENT.mode()[0])

In [ ]:
previous_data.AMT_DOWN_PAYMENT.isnull().sum()

0

In [ ]:
previous_data.AMT_ANNUITY.isnull().sum()

0

Finding outliers for previous application dataset

In [ ]:
# Checking the counts of different contract status
previous_data.NAME_CONTRACT_STATUS.value_counts()

Approved        1036781
Canceled         316319
Refused          290678
Unused offer      26436
Name: NAME_CONTRACT_STATUS, dtype: int64

In [ ]:
# checking counts for payment type
previous_data.NAME_PAYMENT_TYPE.value_counts()

Cash through the bank                        1033552
XNA                                           627384
Non-cash from your account                      8193
Cashless from the account of the employer       1085
Name: NAME_PAYMENT_TYPE, dtype: int64

Bivariate Analysis

Cash through the bank is the most common type of payment method for loan applications.

Merging the two dataframes ( current application with previous application ) :

In [ ]:
#Merging application dataset with previous application dataset
df_merged=pd.merge(left = applied_update, right = previous_data, how='inner', on = 'SK_ID_CURR', suffixes='_x')
df_merged.columns

<ipython-input-73-099a695848cf>:2: FutureWarning: Passing 'suffixes' as a <class 'str'>, is not supported and may give unexpected results. Provide 'suffixes' as a tuple instead. In the future a 'TypeError' will be raised.
  df_merged=pd.merge(left = applied_update, right = previous_data, how='inner', on = 'SK_ID_CURR', suffixes='_x')


Index(['SK_ID_CURR', 'TARGET', 'NAME_CONTRACT_TYPE_', 'CODE_GENDER',
       'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'CNT_CHILDREN', 'AMT_INCOME_TOTAL',
       'AMT_CREDIT_', 'AMT_ANNUITY_', 'AMT_GOODS_PRICE', 'NAME_TYPE_SUITE',
       'NAME_INCOME_TYPE', 'NAME_EDUCATION_TYPE', 'NAME_FAMILY_STATUS',
       'NAME_HOUSING_TYPE', 'REGION_POPULATION_RELATIVE', 'DAYS_BIRTH',
       'DAYS_EMPLOYED', 'DAYS_REGISTRATION', 'DAYS_ID_PUBLISH',
       'OCCUPATION_TYPE', 'CNT_FAM_MEMBERS', 'WEEKDAY_APPR_PROCESS_START_',
       'HOUR_APPR_PROCESS_START_', 'REG_REGION_NOT_LIVE_REGION',
       'REG_REGION_NOT_WORK_REGION', 'LIVE_REGION_NOT_WORK_REGION',
       'REG_CITY_NOT_LIVE_CITY', 'REG_CITY_NOT_WORK_CITY',
       'LIVE_CITY_NOT_WORK_CITY', 'ORGANIZATION_TYPE', 'EXT_SOURCE_2',
       'EXT_SOURCE_3', 'OBS_30_CNT_SOCIAL_CIRCLE', 'DEF_30_CNT_SOCIAL_CIRCLE',
       'OBS_60_CNT_SOCIAL_CIRCLE', 'DEF_60_CNT_SOCIAL_CIRCLE',
       'DAYS_LAST_PHONE_CHANGE', 'AMT_REQ_CREDIT_BUREAU_HOUR',
       'AMT_REQ_CREDIT_BU

In [ ]:
df_merged.shape

(1413701, 71)

In [ ]:
df_merged.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE_ CODE_GENDER  FLAG_OWN_CAR  \
0      100002       1          Cash loans           M             0   
1      100003       0          Cash loans           F             0   
2      100003       0          Cash loans           F             0   
3      100003       0          Cash loans           F             0   
4      100004       0     Revolving loans           M             1   

   FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT_  AMT_ANNUITY_  \
0                1             0          202500.0     406597.5       24700.5   
1                0             0          270000.0    1293502.5       35698.5   
2                0             0          270000.0    1293502.5       35698.5   
3                0             0          270000.0    1293502.5       35698.5   
4                1             0           67500.0     135000.0        6750.0   

   ...  CODE_REJECT_REASON NAME_CLIENT_TYPE   NAME_GOODS_CATEGORY  \
0  ...                 XAP              New              Vehicles   
1  ...                 XAP         Repeater                   XNA   
2  ...                 XAP        Refreshed             Furniture   
3  ...                 XAP        Refreshed  Consumer Electronics   
4  ...                 XAP              New                Mobile   

  NAME_PORTFOLIO NAME_PRODUCT_TYPE             CHANNEL_TYPE  SELLERPLACE_AREA  \
0            POS               XNA                    Stone               500   
1           Cash            x-sell  Credit and cash offices                -1   
2            POS               XNA                    Stone              1400   
3            POS               XNA             Country-wide               200   
4            POS               XNA         Regional / Local                30   

   NAME_SELLER_INDUSTRY  NAME_YIELD_GROUP          PRODUCT_COMBINATION  
0       Auto technology        low_normal      POS other with interest  
1                   XNA        low_normal             Cash X-Sell: low  
2             Furniture            middle   POS industry with interest  
3  Consumer electronics            middle  POS household with interest  
4          Connectivity            middle  POS mobile without interest  

[5 rows x 71 columns]

In [ ]:
#Dividing the data by NAME_CONTRACT_STATUS into separate data frames for better clarity 
merge_app_Refused = df_merged[df_merged["NAME_CONTRACT_STATUS"]  == 'Refused']
merge_app_Approved = df_merged[df_merged["NAME_CONTRACT_STATUS"]  == 'Approved']
merge_app_Canceled = df_merged[df_merged["NAME_CONTRACT_STATUS"]  == 'Canceled']
merge_app_Unused = df_merged[df_merged["NAME_CONTRACT_STATUS"]  == 'Unused offer']

In [ ]:
print(merge_app_Refused.shape)
print(merge_app_Approved.shape)
print(merge_app_Canceled.shape)
print(merge_app_Unused.shape)

(245390, 71)
(886099, 71)
(259441, 71)
(22771, 71)


In [ ]:
df_merged.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE_ CODE_GENDER  FLAG_OWN_CAR  \
0      100002       1          Cash loans           M             0   
1      100003       0          Cash loans           F             0   
2      100003       0          Cash loans           F             0   
3      100003       0          Cash loans           F             0   
4      100004       0     Revolving loans           M             1   

   FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT_  AMT_ANNUITY_  \
0                1             0          202500.0     406597.5       24700.5   
1                0             0          270000.0    1293502.5       35698.5   
2                0             0          270000.0    1293502.5       35698.5   
3                0             0          270000.0    1293502.5       35698.5   
4                1             0           67500.0     135000.0        6750.0   

   ...  CODE_REJECT_REASON NAME_CLIENT_TYPE   NAME_GOODS_CATEGORY  \
0  ...                 XAP              New              Vehicles   
1  ...                 XAP         Repeater                   XNA   
2  ...                 XAP        Refreshed             Furniture   
3  ...                 XAP        Refreshed  Consumer Electronics   
4  ...                 XAP              New                Mobile   

  NAME_PORTFOLIO NAME_PRODUCT_TYPE             CHANNEL_TYPE  SELLERPLACE_AREA  \
0            POS               XNA                    Stone               500   
1           Cash            x-sell  Credit and cash offices                -1   
2            POS               XNA                    Stone              1400   
3            POS               XNA             Country-wide               200   
4            POS               XNA         Regional / Local                30   

   NAME_SELLER_INDUSTRY  NAME_YIELD_GROUP          PRODUCT_COMBINATION  
0       Auto technology        low_normal      POS other with interest  
1                   XNA        low_normal             Cash X-Sell: low  
2             Furniture            middle   POS industry with interest  
3  Consumer electronics            middle  POS household with interest  
4          Connectivity            middle  POS mobile without interest  

[5 rows x 71 columns]

In [ ]:
df_merged.isnull().sum().head(60)

SK_ID_CURR                          0
TARGET                              0
NAME_CONTRACT_TYPE_                 0
CODE_GENDER                         0
FLAG_OWN_CAR                        0
FLAG_OWN_REALTY                     0
CNT_CHILDREN                        0
AMT_INCOME_TOTAL                    0
AMT_CREDIT_                         0
AMT_ANNUITY_                       93
AMT_GOODS_PRICE                  1208
NAME_TYPE_SUITE                     0
NAME_INCOME_TYPE                    0
NAME_EDUCATION_TYPE                 0
NAME_FAMILY_STATUS                  0
NAME_HOUSING_TYPE                   0
REGION_POPULATION_RELATIVE          0
DAYS_BIRTH                          0
DAYS_EMPLOYED                  273583
DAYS_REGISTRATION                   0
DAYS_ID_PUBLISH                     0
OCCUPATION_TYPE                     0
CNT_FAM_MEMBERS                     0
WEEKDAY_APPR_PROCESS_START_         0
HOUR_APPR_PROCESS_START_            0
REG_REGION_NOT_LIVE_REGION          0
REG_REGION_N

In [ ]:
(df_merged.isnull().sum()/len(df_merged)*100).sort_values(ascending=False).head(50)

DAYS_EMPLOYED                  19.352253
YEARS_EMPLOYED                 19.352253
AMT_GOODS_PRICE                 0.085449
PRODUCT_COMBINATION             0.022140
AMT_ANNUITY_                    0.006578
AMT_CREDITx                     0.000071
SELLERPLACE_AREA                0.000000
DAYS_BIRTH_BINS                 0.000000
AMT_APPLICATION                 0.000000
AMT_ANNUITYx                    0.000000
NAME_CONTRACT_TYPEx             0.000000
SK_ID_PREV                      0.000000
NAME_YIELD_GROUP                0.000000
AMT_REQ_CREDIT_BUREAU_YEAR      0.000000
WEEKDAY_APPR_PROCESS_STARTx     0.000000
AMT_REQ_CREDIT_BUREAU_QRT       0.000000
AMT_REQ_CREDIT_BUREAU_MON       0.000000
AMT_REQ_CREDIT_BUREAU_WEEK      0.000000
AMT_REQ_CREDIT_BUREAU_DAY       0.000000
AMT_REQ_CREDIT_BUREAU_HOUR      0.000000
AMT_DOWN_PAYMENT                0.000000
HOUR_APPR_PROCESS_STARTx        0.000000
CHANNEL_TYPE                    0.000000
FLAG_LAST_APPL_PER_CONTRACT     0.000000
DEF_60_CNT_SOCIA

In [ ]:
df_merged.DAYS_EMPLOYED.fillna(df_merged.DAYS_EMPLOYED .median() , inplace = True)
df_merged.YEARS_EMPLOYED.fillna(df_merged.YEARS_EMPLOYED.median() , inplace = True) 
df_merged.AMT_GOODS_PRICE .fillna(df_merged.AMT_GOODS_PRICE  .mean() , inplace = True)
df_merged.PRODUCT_COMBINATION .fillna(df_merged.PRODUCT_COMBINATION .mode() , inplace = True) 



In [ ]:
(df_merged.isnull().sum()/len(df_merged)*100).sort_values(ascending=False).head(5)

PRODUCT_COMBINATION           0.022140
AMT_ANNUITY_                  0.006578
AMT_CREDITx                   0.000071
DAYS_LAST_PHONE_CHANGE        0.000000
AMT_REQ_CREDIT_BUREAU_HOUR    0.000000
dtype: float64

In [ ]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
df_merged['NAME_CONTRACT_TYPE_'] = le.fit_transform(df_merged['NAME_CONTRACT_TYPE_'])
df_merged['CODE_GENDER'] = le.fit_transform(df_merged['CODE_GENDER'])
df_merged['NAME_TYPE_SUITE'] = le.fit_transform(df_merged['NAME_TYPE_SUITE'].astype(str))
df_merged['NAME_INCOME_TYPE'] = le.fit_transform(df_merged['NAME_INCOME_TYPE'])
df_merged['NAME_EDUCATION_TYPE'] = le.fit_transform(df_merged['NAME_EDUCATION_TYPE'])
df_merged['NAME_FAMILY_STATUS'] = le.fit_transform(df_merged['NAME_FAMILY_STATUS'])
df_merged['NAME_HOUSING_TYPE'] = le.fit_transform(df_merged['NAME_HOUSING_TYPE'])
df_merged['OCCUPATION_TYPE'] = le.fit_transform(df_merged['OCCUPATION_TYPE'].astype(str))
df_merged['WEEKDAY_APPR_PROCESS_START_'] = le.fit_transform(df_merged['WEEKDAY_APPR_PROCESS_START_'])
df_merged['ORGANIZATION_TYPE'] = le.fit_transform(df_merged['ORGANIZATION_TYPE'])
df_merged['DAYS_BIRTH_BINS'] = le.fit_transform(df_merged['DAYS_BIRTH_BINS'])
df_merged['FLAG_LAST_APPL_PER_CONTRACT'] = le.fit_transform(df_merged['FLAG_LAST_APPL_PER_CONTRACT'])
df_merged['NAME_CASH_LOAN_PURPOSE'] = le.fit_transform(df_merged['NAME_CASH_LOAN_PURPOSE'])
df_merged['NAME_CONTRACT_STATUS'] = le.fit_transform(df_merged['NAME_CONTRACT_STATUS'])
df_merged['NAME_PAYMENT_TYPE'] = le.fit_transform(df_merged['NAME_PAYMENT_TYPE'])
df_merged['CODE_REJECT_REASON'] = le.fit_transform(df_merged['CODE_REJECT_REASON'])
df_merged['NAME_CLIENT_TYPE'] = le.fit_transform(df_merged['NAME_CLIENT_TYPE'])
df_merged['NAME_GOODS_CATEGORY'] = le.fit_transform(df_merged['NAME_GOODS_CATEGORY'])
df_merged['NAME_PORTFOLIO'] = le.fit_transform(df_merged['NAME_PORTFOLIO'])
df_merged['NAME_PRODUCT_TYPE'] = le.fit_transform(df_merged['NAME_PRODUCT_TYPE'])
df_merged['CHANNEL_TYPE'] = le.fit_transform(df_merged['CHANNEL_TYPE'])
df_merged['NAME_SELLER_INDUSTRY'] = le.fit_transform(df_merged['NAME_SELLER_INDUSTRY'])
df_merged['NAME_YIELD_GROUP'] = le.fit_transform(df_merged['NAME_YIELD_GROUP'])
df_merged['PRODUCT_COMBINATION'] = le.fit_transform(df_merged['PRODUCT_COMBINATION'])

In [ ]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1413701 entries, 0 to 1413700
Data columns (total 71 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   SK_ID_CURR                   1413701 non-null  int64  
 1   TARGET                       1413701 non-null  int64  
 2   NAME_CONTRACT_TYPE_          1413701 non-null  int64  
 3   CODE_GENDER                  1413701 non-null  int64  
 4   FLAG_OWN_CAR                 1413701 non-null  int64  
 5   FLAG_OWN_REALTY              1413701 non-null  int64  
 6   CNT_CHILDREN                 1413701 non-null  int64  
 7   AMT_INCOME_TOTAL             1413701 non-null  float64
 8   AMT_CREDIT_                  1413701 non-null  float64
 9   AMT_ANNUITY_                 1413608 non-null  float64
 10  AMT_GOODS_PRICE              1413701 non-null  float64
 11  NAME_TYPE_SUITE              1413701 non-null  int64  
 12  NAME_INCOME_TYPE             1413701 non-n

In [ ]:
X = df_merged.drop('TARGET', axis = 1).values
y = df_merged['TARGET'].values

In [ ]:
#Users need to enter the splitting factor by which dataset should be divided into train and test.
print("Enter the splitting factor (i.e) ratio between train and test")
s_f = float(input())

Enter the splitting factor (i.e) ratio between train and test
0.9


In [ ]:
import math
n_train = math.floor(s_f * X.shape[0])
n_test = math.ceil((1-s_f) * X.shape[0])
X_train = X[:n_train]
y_train = y[:n_train]
X_test = X[n_train:]
y_test = y[n_train:]
print("Total Number of rows in train:",X_train.shape[0])
print("Total Number of rows in test:",X_test.shape[0])

Total Number of rows in train: 1272330
Total Number of rows in test: 141371


In [ ]:
X_train = np.insert(X_train, 2, y_train, axis=1) #Training data should contain both features and target

In [ ]:
df_merged['NAME_CONTRACT_TYPEx'] = le.fit_transform(df_merged['NAME_CONTRACT_TYPEx'])
df_merged['WEEKDAY_APPR_PROCESS_STARTx'] = le.fit_transform(df_merged['WEEKDAY_APPR_PROCESS_STARTx'])

In [ ]:
df_merged.drop('AMT_CREDITx',axis=1,inplace=True)

In [ ]:
df_merged.isna().sum().head(60)

SK_ID_CURR                      0
TARGET                          0
NAME_CONTRACT_TYPE_             0
CODE_GENDER                     0
FLAG_OWN_CAR                    0
FLAG_OWN_REALTY                 0
CNT_CHILDREN                    0
AMT_INCOME_TOTAL                0
AMT_CREDIT_                     0
AMT_ANNUITY_                   93
AMT_GOODS_PRICE                 0
NAME_TYPE_SUITE                 0
NAME_INCOME_TYPE                0
NAME_EDUCATION_TYPE             0
NAME_FAMILY_STATUS              0
NAME_HOUSING_TYPE               0
REGION_POPULATION_RELATIVE      0
DAYS_BIRTH                      0
DAYS_EMPLOYED                   0
DAYS_REGISTRATION               0
DAYS_ID_PUBLISH                 0
OCCUPATION_TYPE                 0
CNT_FAM_MEMBERS                 0
WEEKDAY_APPR_PROCESS_START_     0
HOUR_APPR_PROCESS_START_        0
REG_REGION_NOT_LIVE_REGION      0
REG_REGION_NOT_WORK_REGION      0
LIVE_REGION_NOT_WORK_REGION     0
REG_CITY_NOT_LIVE_CITY          0
REG_CITY_NOT_W

In [ ]:
df_merged.drop('AMT_ANNUITY_',axis=1,inplace=True)

In [ ]:
# Import necessary libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# Split the dataset into training and testing sets
X = df_merged.drop('TARGET', axis=1).values
y = df_merged['TARGET'].values
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
# Train the KNN classifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

KNeighborsClassifier()

In [ ]:
# Predict the target values for the test data
y_pred = knn.predict(X_test)

In [ ]:
# Evaluate the accuracy of the model
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy: ", accuracy)

Accuracy:  0.9078237680421304


In [ ]:
# Evaluate the confusion matrix of the model
confusion_mat = confusion_matrix(y_test, y_pred)
print("Confusion Matrix: \n", confusion_mat)

Confusion Matrix: 
 [[256202   2160]
 [ 23902    477]]
